bundles/org.openhab.binding.hdpowerview/src/main/java/org/openhab/binding/hdpowerview/internal/handler/HDPowerViewShadeHandler.java

```diff
@@ -596,4 +611,100 @@ private void doRefreshShade(RefreshKind kind) {
+        }
+    }
+
+    /**
+     * Helper to do the actual work of initializing the dynamic channels if the respective device supports them.
+     *
+     * @return true if all channels were successfully initialised.
+     */
+    private boolean dynamicChannelsInitializer() {
+        final Capabilities capabilities = this.capabilities;
+        if (capabilities == null) {
+            // stuff not initialised
+            return false;
+        }
+
+        // note: this is an immutable list
+        List<Channel> channels = thing.getChannels();
+        final Stream<Channel> channelStream = channels.stream();
+
+        // predicate to filter the vane channel
+        final Predicate<Channel> vaneChannelFilter = c -> HDPowerViewBindingConstants.CHANNEL_TYPE_VANE
+                .equals(c.getChannelTypeUID());
+
+        // predicate to filter the secondary channel
+        final Predicate<Channel> secChannelFilter = c -> HDPowerViewBindingConstants.CHANNEL_TYPE_SECONDARY
+                .equals(c.getChannelTypeUID());
+
+        // current and required state of the vane channel
+        final boolean vaneChannelExisting = channelStream.anyMatch(vaneChannelFilter);
+        final boolean vaneChannelRequired = capabilities.usesVaneChannel();
+
+        final boolean secChannelExisting = channelStream.anyMatch(secChannelFilter);
+        final boolean secChannelRequired = capabilities.usesSecondaryChannel();
+
+        if ((vaneChannelExisting == vaneChannelRequired) && (secChannelExisting == secChannelRequired)) {
+            // nothing to do
+            return true;
+        }
+
+        // make a mutable copy of the original immutable channel list
+        channels = new ArrayList<>(channels);
+
+        boolean error = false;
+        if (vaneChannelRequired) {
+            // build the vane channel
+            Channel vaneChannel = ChannelBuilder
+                    .create(new ChannelUID(thing.getUID(), HDPowerViewBindingConstants.CHANNEL_SHADE_VANE))
+                    .withType(HDPowerViewBindingConstants.CHANNEL_TYPE_VANE).withKind(ChannelKind.STATE)
+                    .withAcceptedItemType(CoreItemFactory.DIMMER)
+                    .withDescription(translater.getText("channel-type.hdpowerview.shade-vane.description"))
+                    .withLabel(translater.getText("channel-type.hdpowerview.shade-vane.label")).build();
+
+            // add the vane channel
+            error |= !channels.add(vaneChannel);
+        } else {
+            // remove the vane channel
+            error |= !channels.removeIf(vaneChannelFilter);
+        }
+
+        if (secChannelRequired) {
+            // build the secondary channel
+            Channel secChannel = ChannelBuilder
+                    .create(new ChannelUID(thing.getUID(),
+                            HDPowerViewBindingConstants.CHANNEL_SHADE_SECONDARY_POSITION))
+                    .withType(HDPowerViewBindingConstants.CHANNEL_TYPE_SECONDARY).withKind(ChannelKind.STATE)
```

Comment:
Why not use channel type `shade-position` like the static channel did? We now have redundant channel types only different in label and description, and this is set explicitly anyway.

Reply:
Makes sense.

Reply:
Reopening this since unresolved. Proposition was to have only one channel type for shade position which can be used for both channels - like previously. They only differ in name, so there is no reason for duplication.

Reply:
I want to follow a consistent approach towards the creation of the three dynamic channels (position, vane and secondary). And the problem with your approach is due to the internationalisation, where one would be creating the position and vane channels using the internationalisation properties key for the channel-type and the secondary channel using the internationalisation properties key for the thing-channel. This is messy. So I want to create all three things, and internationalize them using the exact same algorithm.

PS your statement _"They only differ in name"_ is in fact wrong -- in fact they differ in name, label and description.

Reply:
You are of course right, label and description - I'm considering them a pair. :-) There is no such thing as name, but of course they also have their own id. My argument was that from a **design**/model perspective it's in fact messy to duplicate channel types. It seems your reasoning for doing so was **implementation** issues. Knowing this, I'll have another look and see if I can come up with something that would be clean and still avoid the redundancy.